The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9642665526/job/26591034145) [(plain)](https://github.com/rust-lang-ci/rust/commit/4f44f046d29aa1b6c8ee15b0860df01199f3daab/checks/26591034145/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  99% (47788/48266)
Updating files:  99% (48156/48266)
Updating files: 100% (48266/48266)
Updating files: 100% (48266/48266), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'4f44f046d29aa1b6c8ee15b0860df01199f3daab'
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
[RUSTC-TIMING] run_make_support test:false 3.523
    Finished `release` profile [optimized] target(s) in 40.63s
[TIMING] core::build_steps::test::RunMakeSupport { compiler: Compiler { stage: 2, host: x86_64-pc-windows-msvc }, target: x86_64-pc-windows-msvc } -- 40.654
##[group]Testing stage2 compiletest suite=run-make mode=run-make (x86_64-pc-windows-msvc)
error: detected unknown compiletest test directive `needs-ar` in C:\a\rust\rust\tests\run-make\invalid-library\rmake.rs:1
errors encountered during EarlyProps parsing: C:\a\rust\rust\tests\run-make\invalid-library\rmake.rs
errors encountered during EarlyProps parsing
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 1:23:55
Build completed unsuccessfully in 1:23:55
make: *** [Makefile:102: ci-msvc-ps1] Error 1
  network time: Mon, 24 Jun 2024 10:45:41 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

--- a/src/tools/run-make-support/Cargo.toml
+++ b/src/tools/run-make-support/Cargo.toml
@@ -9,3 +9,4 @@ similar = "2.5.0"
 wasmparser = "0.118.2"
 regex = "1.8" # 1.8 to avoid memchr 2.6.0, as 2.5.0 is pinned in the workspace
 gimli = "0.28.1"
+ar = "0.9.0"

--- a/src/tools/run-make-support/src/lib.rs
+++ b/src/tools/run-make-support/src/lib.rs
@@ -61,6 +61,14 @@ pub fn target() -> String {
     env_var("TARGET")
 }
 
+/// `AR`
+#[track_caller]
+#[must_use]
+pub fn ar_command() -> Command {
+    let ar_path = env_var("AR");
+    Command::new(ar_path)
+}
+
 /// Check if target is windows-like.
 #[must_use]
 pub fn is_windows() -> bool {

--- a/src/tools/run-make-support/src/lib.rs
+++ b/src/tools/run-make-support/src/lib.rs
@@ -63,10 +63,15 @@ pub fn target() -> String {
 
 /// `AR`
 #[track_caller]
-#[must_use]
-pub fn ar_command() -> Command {
-    let ar_path = env_var("AR");
-    Command::new(ar_path)
+pub fn ar(inputs: &[impl AsRef<Path>], output_path: impl AsRef<Path>) {
+    let output = fs::File::create(&output_path).expect(&format!(
+        "the file in path \"{}\" could not be created",
+        output_path.as_ref().display()
+    ));
+    let mut builder = ar::Builder::new(output);
+    for input in inputs {
+        builder.append_path(input).unwrap();
+    }
 }
 
 /// Check if target is windows-like.

--- a/src/tools/tidy/src/allowed_run_make_makefiles.txt
+++ b/src/tools/tidy/src/allowed_run_make_makefiles.txt
@@ -54,7 +54,6 @@ run-make/incr-add-rust-src-component/Makefile
 run-make/incr-foreign-head-span/Makefile
 run-make/interdependent-c-libraries/Makefile
 run-make/intrinsic-unreachable/Makefile
-run-make/invalid-library/Makefile
 run-make/issue-107094/Makefile
 run-make/issue-109934-lto-debuginfo/Makefile
 run-make/issue-14698/Makefile

--- a/tests/run-make/invalid-library/Makefile
+++ b/tests/run-make/invalid-library/Makefile
@@ -1,6 +1,0 @@
-include ../tools.mk
-
-all:
-	touch $(TMPDIR)/lib.rmeta
-	$(AR) crus $(TMPDIR)/libfoo-ffffffff-1.0.rlib $(TMPDIR)/lib.rmeta
-	$(RUSTC) foo.rs 2>&1 | $(CGREP) "found invalid metadata"

--- a/tests/run-make/invalid-library/rmake.rs
+++ b/tests/run-make/invalid-library/rmake.rs
@@ -1,0 +1,8 @@
+use run_make_support::fs_wrapper::create_file;
+use run_make_support::{ar_command, rustc};
+
+fn main() {
+    create_file("lib.rmeta");
+    ar_command().arg("crus").arg("libfoo-ffffffff-1.0.rlib").arg("lib.rmeta").run();
+    rustc().input("foo.rs").run_fail().assert_stderr_contains("found invalid metadata");
+}

--- a/tests/run-make/invalid-library/rmake.rs
+++ b/tests/run-make/invalid-library/rmake.rs
@@ -1,8 +1,8 @@
 use run_make_support::fs_wrapper::create_file;
-use run_make_support::{ar_command, rustc};
+use run_make_support::{ar, rustc};
 
 fn main() {
     create_file("lib.rmeta");
-    ar_command().arg("crus").arg("libfoo-ffffffff-1.0.rlib").arg("lib.rmeta").run();
+    ar(&["lib.rmeta"], "libfoo-ffffffff-1.0.rlib");
     rustc().input("foo.rs").run_fail().assert_stderr_contains("found invalid metadata");
 }

